### PR TITLE
async-function-support

### PR DESCRIPTION
**Pull request by FarukOzderim**
- support async functions with Blocks
- refactor app a bit

Related: #727 

This took some time to correctly work on Interface, it seems lambda functions are always wicked :D

### CODE REVIEW

**Comment by FarukOzderim** [comment]
You can try these:


```
import asyncio
import gradio as gr


async def async_func(*args):
    await asyncio.sleep(5)
    return "Hello after 5 second"


gr.Interface(async_func, "text", "text").launch()
```


```
import asyncio
import gradio as gr
with gr.Blocks() as demo:
    text = gr.components.Textbox()
    button = gr.components.Button()
    button.click(async_func, text, text)
    demo.launch()

```

File: gradio/routes.py [hunk 1]
```diff
@@ -78,199 +78,228 @@ class PredictBody(BaseModel):
...
-
-    @app.get("/login_check")
-    @app.get("/login_check/")
-    def login_check(user: str = Depends(get_current_user)):
-        if app.auth is None or not (user is None):
-            return
-        raise HTTPException(
-            status_code=status.HTTP_401_UNAUTHORIZED, detail="Not authenticated"
-        )
+class App(FastAPI):
```
Comment (FarukOzderim) [inline gradio/routes.py:81]: No functional change here, just tidied things a little bit into OOP.

**Comment by FarukOzderim** [comment]
There are some problems, not completed yet.

File: gradio/routes.py (hunk 1 above)
Comment (abidlabs) [reply gradio/routes.py:81]: Nice, thanks!

**Review by abidlabs — APPROVED** [review]

**Comment by FarukOzderim** [comment]
> Looking great @FarukOzderim!
> 
> I would only that we add a test in `test_blocks` with an async function to ensure everything works as expected.

Great idea!

**Pull request merged**